File: Android/PlayerProj/animplayer/src/main/java/com/tencent/qgame/animplayer/mask/MaskRender.kt
/*
 * Tencent is pleased to support the open source community by making vap available.
 *
 * Copyright (C) 2020 THL A29 Limited, a Tencent company.  All rights reserved.
 *
 * Licensed under the MIT License (the "License"); you may not use this file except in
 * compliance with the License. You may obtain a copy of the License at
 *
 * http://opensource.org/licenses/MIT
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License is
 * distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied. See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.tencent.qgame.animplayer.mask

import android.opengl.GLES11Ext
import android.opengl.GLES20
import com.tencent.qgame.animplayer.AnimConfig
import com.tencent.qgame.animplayer.Constant
import com.tencent.qgame.animplayer.PointRect
import com.tencent.qgame.animplayer.RefVec2
import com.tencent.qgame.animplayer.util.GlFloatArray
import com.tencent.qgame.animplayer.util.TexCoordsUtil
import com.tencent.qgame.animplayer.util.VertexUtil

/**
 * vapx 渲染
 */
class MaskRender(private val maskAnimPlugin: MaskAnimPlugin) {
    companion object {
        private const val TAG = "${Constant.TAG}.MaskRender"
    }

    var maskShader: MaskShader? = null
    var vertexArray = GlFloatArray()
    private var maskArray = GlFloatArray()
    /**
     * shader 与 texture初始化
     */
    fun initMaskShader(edgeBlur: Boolean) {
        // shader 初始化
        maskShader = MaskShader(edgeBlur)
        GLES20.glDisable(GLES20.GL_DEPTH_TEST) // 关闭深度测试
    }

    fun renderFrame(config: AnimConfig) {
        val videoTextureId = maskAnimPlugin.player.decoder?.render?.getExternalTexture() ?: return
        if (videoTextureId <= 0) return
        val shader = this.maskShader ?: return
        var maskTexId: Int = config.maskConfig?.getMaskTexId() ?: return
        val maskBitmap = config.maskConfig?.alphaMaskBitmap ?: return
        val maskTexRect = config.maskConfig?.maskTexPair?.first ?: return
        val maskTexRefVec2 = config.maskConfig?.maskTexPair?.second ?: return
        val maskPositionRect = config.maskConfig?.maskPositionPair?.first ?: PointRect(
            0,
            0,
            config.width,
            config.height
        )
        val maskPositionRefVec2 =
            config.maskConfig?.maskPositionPair?.second ?: RefVec2(config.width, config.height)
        shader.useProgram()
        // 顶点坐标
        vertexArray.setArray(
            VertexUtil.create(
                maskPositionRefVec2.w,
                maskPositionRefVec2.h,
                maskPositionRect,
                vertexArray.array
            )
        )
        vertexArray.setVertexAttribPointer(shader.aPositionLocation)

        if (maskTexId <= 0 && !maskBitmap.isRecycled) {
            maskTexId = config.maskConfig?.updateMaskTex() ?: 0
        }
        if (maskTexId > 0) {
            maskArray.setArray(
                TexCoordsUtil.create(
                    maskTexRefVec2.w,
                    maskTexRefVec2.h,
                    maskTexRect,
                    maskArray.array
                )
            )
            maskArray.setVertexAttribPointer(shader.aTextureMaskCoordinatesLocation)
            // 绑定alpha纹理
            GLES20.glActiveTexture(GLES20.GL_TEXTURE0)
            GLES20.glBindTexture(GLES20.GL_TEXTURE_2D, maskTexId)
            GLES20.glTexParameterf(
                GLES20.GL_TEXTURE_2D,
                GLES20.GL_TEXTURE_MIN_FILTER,
                GLES20.GL_NEAREST.toFloat()
            )
            GLES20.glTexParameterf(
                GLES20.GL_TEXTURE_2D,
                GLES20.GL_TEXTURE_MAG_FILTER,
                GLES20.GL_LINEAR.toFloat()
            )
            GLES20.glTexParameteri(
                GLES11Ext.GL_TEXTURE_EXTERNAL_OES,
                GLES20.GL_TEXTURE_WRAP_S,
                GLES20.GL_CLAMP_TO_EDGE
            )
            GLES20.glTexParameteri(
                GLES11Ext.GL_TEXTURE_EXTERNAL_OES,
                GLES20.GL_TEXTURE_WRAP_T,
                GLES20.GL_CLAMP_TO_EDGE
            )
            GLES20.glUniform1i(shader.uTextureMaskUnitLocation, 0)

            GLES20.glEnable(GLES20.GL_BLEND)
            // 基于源象素alpha通道值的半透明混合函数
            GLES20.glBlendFuncSeparate(
                GLES20.GL_ONE,
                GLES20.GL_SRC_ALPHA,
                GLES20.GL_ZERO,
                GLES20.GL_SRC_ALPHA
            )
            // draw
            GLES20.glDrawArrays(GLES20.GL_TRIANGLE_STRIP, 0, 4)

            GLES20.glDisable(GLES20.GL_BLEND)
        }

    }

}

File: ohos/vap_module/src/main/cpp/mask/mask_render.cpp
/*
 * Copyright (C) 2024 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "mask_render.h"

#include "log.h"

#include <GLES2/gl2ext.h>
#include "common_const.h"

using namespace CommonConst;

void MaskRender::RenderFrame(MaskConfig maskConfig)
{
    auto maskTexId = maskConfig.GetMaskTexId();
    if (maskTexId <= 0) {
        LOGE("maskTexId not ready, cont Rendered");
    }
    
    auto texTuple = maskConfig.maskTexPair_;
    auto maskTexRect = std::get<0>(texTuple);
    auto maskTexRefVec2 = std::get<1>(texTuple);
    
    auto positionTuple = maskConfig.maskPositionPair_;
    PointRect maskPositionRect;
    if (std::get<0>(positionTuple).w != 0 && std::get<0>(positionTuple).h != 0) {
        maskPositionRect = std::get<0>(positionTuple);
    } else {
        maskPositionRect = PointRect(0, 0, maskConfig.width_, maskConfig.height_);
    }

    RefVec2 maskPositionRefVec2;
    if (std::get<1>(positionTuple).w != 0 && std::get<1>(positionTuple).h !=0) {
        maskPositionRefVec2 = std::get<1>(positionTuple);
    } else {
        maskPositionRefVec2 = RefVec2(maskConfig.width_, maskConfig.height_);
    }
    
    m_maskShader->UseProgram();
    m_vertexArray.Create(maskPositionRefVec2.w, maskPositionRefVec2.h, maskPositionRect, m_vertexArray.array.get());
    m_vertexArray.SetArray(m_vertexArray.array.get());
    m_vertexArray.EnableVertexAttrib(m_maskShader->m_aPositionLocation);
    
    m_maskArray.Create(
        maskTexRefVec2.w,
        maskTexRefVec2.h,
        maskTexRect,
        m_maskArray.array.get()
    );
    m_maskArray.SetArray(m_maskArray.array.get());
    m_maskArray.EnableVertexAttrib(m_maskShader->m_aTextureMaskCoordinatesLocation);
    glActiveTexture(GL_TEXTURE0);
    glBindTexture(GL_TEXTURE_2D, maskTexId);
    glTexParameterf(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
    glTexParameterf(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
    glUniform1i(m_maskShader->m_uTextureMaskUnitLocation, 0);

    glEnable(GL_BLEND);
    glBlendFuncSeparate(GL_ONE, GL_SRC_ALPHA, GL_ZERO, GL_SRC_ALPHA);
    glDrawArrays(GL_TRIANGLE_STRIP, 0, FOUR);
    glDisable(GL_BLEND);
}

MaskRender::MaskRender(bool edgeBlur)
{
    m_maskShader = std::make_unique<class MaskShader>(edgeBlur);
    glDisable(GL_DEPTH_TEST);
}